src/engine/index.ts

```diff
@@ -110,27 +111,26 @@ class Engine {
   }
 
   /** Merge a cast into the set */
-  mergeCast(cast: Cast): Result<void, string> {
+  async mergeCast(cast: Cast): Promise<Result<void, string>> {
     try {
       const username = cast.data.username;
 
       const signerChanges = this._users.get(username);
       if (!signerChanges) {
         return err('mergeCast: unknown user');
       }
-
-      if (!this.validateMessage(cast).isOk()) {
-        return this.validateMessage(cast);
+      if (!(await this.validateMessage(cast)).isOk()) {
```

Comment:
this is a bit hard to read since there are so many clauses, i would recommend splitting into a separate line and using the isErr() method to avoid the negation.  

```suggestion
     const isCastValidResult = await this.validateMessage(cast); 
      if (isCastValidResult.isErr()) {
```